src/TeleportJoin.sol

```diff
@@ -253,21 +293,116 @@ contract TeleportJoin {
     /**
     * @dev External function that repays debt with DAI previously pushed to this contract (in general coming from the bridges)
     * @param sourceDomain domain where the DAI is coming from
-    * @param batchedDaiToFlush Amount of DAI that is being processed for repayment
+    * @param targetDomain this domain
+    * @param amount Amount of DAI that is being processed for repayment
     **/
-    function settle(bytes32 sourceDomain, uint256 batchedDaiToFlush) external {
-        require(batchedDaiToFlush <= 2 ** 255, "TeleportJoin/overflow");
-        daiJoin.join(address(this), batchedDaiToFlush);
+    function settle(bytes32 sourceDomain, bytes32 targetDomain, uint256 amount) external {
+        require(targetDomain == domain, "TeleportJoin/incorrect-targetDomain");
+        require(amount <= 2 ** 255, "TeleportJoin/overflow");
+        dai.transferFrom(msg.sender, address(this), amount);
+        daiJoin.join(address(this), amount);
         if (vat.live() == 1) {
             (, uint256 art_) = vat.urns(ilk, address(this)); // rate == RAY => normalized debt == actual debt
-            uint256 amtToPayBack = _min(batchedDaiToFlush, art_);
+            uint256 amtToPayBack = _min(amount, art_);
             vat.frob(ilk, address(this), address(this), address(this), -int256(amtToPayBack), -int256(amtToPayBack));
             vat.slip(ilk, address(this), -int256(amtToPayBack));
             unchecked {
                 art = art_ - amtToPayBack; // Always safe operation
             }
         }
-        debt[sourceDomain] -= int256(batchedDaiToFlush);
-        emit Settle(sourceDomain, batchedDaiToFlush);
+        debt[sourceDomain] -= int256(amount);
+        emit Settle(sourceDomain, amount);
+    }
+
+    /**
+    * @notice Initiate Maker teleport
+    * @dev Will fire a teleport event, burn the dai and initiate a censorship-resistant slow-path message
+    * @param targetDomain The target domain to teleport to
+    * @param receiver The receiver address of the DAI on the target domain
+    * @param amount The amount of DAI to teleport
+    **/
+    function initiateTeleport(
```

Comment:
I know inheritance is frown upon but I wonder if it would be worth splitting this contract into `TeleportInboundGateway` (basically the current TeleportJoin) and `TeleportOutboundGateway` (an abstract contract with just `initiateTeleport` and `flush`). That could make the present contract more readable by keeping independent logic separate.

Reply:
I'm not too picky when it comes to inheritance, so I'm good either way. Let's see what the rest of the team has to say. @gbalabasquer

Reply:
I think this one got outdated now and can be resolved.